packages/core/src/internal/clients/leaderboard/AbstractLeaderboardClient.ts

```diff
@@ -0,0 +1,199 @@
+import {
+  LeaderboardDelete,
+  LeaderboardFetch,
+  LeaderboardGetRank,
+  LeaderboardLength,
+  LeaderboardRemoveElements,
+  LeaderboardUpsert,
+  SortedSetOrder,
+} from '../../../index';
+import {
+  ILeaderboardClient,
+  LeaderboardFetchByRankOptions,
+  LeaderboardFetchByScoreOptions,
+} from '../../../clients/ILeaderboardClient';
+import {InternalLeaderboardClient} from './InternalLeaderboardClient';
+
+export interface BaseLeaderboardClientProps {
+  createLeaderboardClient: () => InternalLeaderboardClient;
+}
+export abstract class AbstractLeaderboardClient implements ILeaderboardClient {
+  // making these protected until we fully abstract away the nodejs client
+  protected readonly leaderboardClient: InternalLeaderboardClient;
+
+  protected constructor(props: BaseLeaderboardClientProps) {
+    this.leaderboardClient = props.createLeaderboardClient();
+  }
+
+  /**
+   * Updates elements in a leaderboard or inserts elements if they do not already exist.
+   * The leaderboard is also created if it does not already exist.
+   *
+   * @param {string} cacheName - The cache containing the leaderboard.
+   * @param {string} leaderboardName - The leaderboard to upsert to.
+   * @param {Map<bigint, number>} elements - The ID->score pairs to add to the leaderboard.
+   * @returns {Promise<LeaderboardUpsert.Response>} -
+   * {@link LeaderboardUpsert.Success} on success.
+   * {@link LeaderboardUpsert.Error} on failure.
+   */
+  public async leaderboardUpsert(
```

Comment:
Have we had conversations with the server team about limits for this API? max number of elements and/or max payload size? If we are doing this as a unary API rather than a streaming one we will need to provide guidance to users around how to batch things.

This comment is not a blocker for this PR but I think we might want to augment this docstring with a link to a docs page or examples of how to use this for bulk ingest, so we should probably ticket that.

Reply:
@cprice404 
The API proposal draft noted that there's an upper limit of 8192 elements for the upsert, remove, fetch by score, and fetch by rank APIs. I added a new validator to check for number of elements per batch and also added that note to the docstrings. I can add another task to the github issue for creating docs/examples and linking that as well.

Reply:
let's not put validation code in the clients for this for now; if they change something on the server we don't want to have to ship a new release of the SDK. docstrings + ticket is fine